modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/PainlessParser.java

```diff
@@ -2834,8 +2837,21 @@ public final LambdaContext lambda() throws RecognitionException {
       }
       setState(448);
       match(ARROW);
-      setState(449);
-      block();
+      setState(451);
+      switch ( getInterpreter().adaptivePredict(_input,38,_ctx) ) {
```

Comment:
> in the braces case, the semicolon is still required!

I can have a look at that after you merge this if you like. The semicolon not required fro braces case is a sneaky lexer trick so it shouldn't have anything to do with the parser....